add arbitrary cross-validation functions

wdearden opened the pull request:
I want to resolve #8. Just to note that this a code-breaking change. I couldn't find a clean way to edit the defaults for `learn_cvpairs` so I changed the default to `crossv_mc` with `n = 1`. Unfortunately, since `crossv_mc` does not have a default for n the code is a little ugly. I also edited the tests and the README to reflect these changes.

By the way, this is my first pull request so if you have any comments or criticisms it is greatly appreciated.

Commit:
add arbitrary cross-validation functions

drsimonj commented on DESCRIPTION:
```diff
@@ -21,8 +21,9 @@ Imports:
 Suggests: testthat,
     knitr,
-    rmarkdown
+    rmarkdown,
```
Is there any new functionality that depends on resamplr? If not, please remove it from this list.

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,37 @@
+#' learn_cvpairs(pl, resamplr::holdout_frac, size = 0.3, K = 5L)
 #' @export
-learn_cvpairs <- function(pl, k = NULL, n = 1, test = 0.2) {
+learn_cvpairs <- function(pl, .f = function(data) modelr::crossv_mc(data, n = 1), ...) {
```
Leave `.f` as an empty argument. Ie. `function(pl, .f, ...)`

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,37 @@
 #'
 #' @inheritParams pipelearner_params
-#' @inheritParams modelr::crossv_mc
+#' @param .f a function which takes a \code{data} argument
```
This needs to be expanded because the output of the function is expected to conform to a very specific structure. Ie, it must take a `data.frame` and return ...

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,37 @@
-
-  pl$cv_pairs <- cv_pairs
+learn_cvpairs.pipelearner <- function(pl, .f = function(data) modelr::crossv_mc(data, n = 1), ...) {
+  pl$cv_pairs <- purrr::invoke(.f, data = pl$data, ...)
```
What is the purpose of `purrr::invoke` instead of just using `.f(data = pl$data, ...)`?

drsimonj replied:
Can cope with default situation here via `if (missing(.f)) {...}`. Must be sure to carefully document this default case.

drsimonj replied:
If we're going to allow any function to be used, there must be a check that the output of the function is valid. 

Function results should be saved as a temporary object, checks should be made that it contains the relevant info (ie. three columns `train`, `test`, and `.id`, which are respectively of type "list", "list", and "character". Would be good to also check that the train and test objects are, or can be coerced, to data frames. Could be lenient with `.id` and allow any discrete options like factors as well.

If checks are passed, object can be attached to `pl` and return `pl`.

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,37 @@
-  }
-
-  pl$cv_pairs <- cv_pairs
+learn_cvpairs.pipelearner <- function(pl, .f = function(data) modelr::crossv_mc(data, n = 1), ...) {
```
The proposed change will make `modelr::crossv_mc` and `modelr::crossv_kfold` be used a lot with pipelearner. As such, please import these functions to make then directly available (without need for `modelr::`). Import should be done in [R/utils.R](https://github.com/drsimonj/pipelearner/blob/master/R/utils.R#L11). Both functions can be imported under the same name.

drsimonj commented on README.Rmd:
````diff
@@ -89,11 +89,11 @@ pl %>% learn()
 ```{r}
 pl %>%
-  learn_cvpairs(k = 5) %>% 
+  learn_cvpairs(resamplr::holdout_n, size = 10, K = 5) %>% 
````
Please stick to the modelr functions in the README which, after importing (see other comment) can be referenced without `modelr::`. You're welcome to create a vignette to expand on this functionality, but README is already a bit too long to bring in novel aspects like this.

drsimonj commented on README.Rmd:
```diff
@@ -166,7 +166,7 @@ After initialization, pipelearner functions can be combined in a single pipeline
 iris %>% 
   pipelearner() %>%
-  learn_cvpairs(n = 50) %>%
+  learn_cvpairs(modelr::crossv_mc, n = 50) %>%
```
Can remove `modelr::` after importing `crossv_mc` (see other comment)

drsimonj commented on README.Rmd:
```diff
@@ -212,7 +212,7 @@ d <- weather %>%
 results <- d %>% 
   pipelearner() %>% 
-  learn_cvpairs(n = 50, test = .15) %>% 
+  learn_cvpairs(modelr::crossv_mc, n = 50, test = .15) %>% 
```
Can remove `modelr::` after importing `crossv_mc` (see other comment)

drsimonj commented on README.Rmd:
```diff
@@ -250,7 +250,7 @@ The example below fits a decision tree and random forest to 20 folds of a subset
 results <-  d %>% 
   pipelearner() %>% 
-  learn_cvpairs(k = 20) %>% 
+  learn_cvpairs(modelr::crossv_kfold, k = 20) %>% 
```
Can remove `modelr::` after importing `crossv_kfold` (see other comment)

drsimonj commented on DESCRIPTION:
```diff
@@ -2,7 +2,7 @@ Package: pipelearner
 Type: Package
 Version: 0.0.0.9000
```
Because the proposed change will break the code, please increment version to 0.0.0.9001

drsimonj commented:
Hi William, thanks for your efforts and patience. I think you've proposed a valuable change to pipelearner. 

Given that it's a reasonably serious change, I expect it to take a few iterations to get right. I've now left comments on your commit code suggesting how it could be changed to better align with the package as a whole and help make everything more generalisable/robust. I may add later, but should be plenty to work on for now.

Don't hesitate to reply if you're unsure of or disagree with anything. Otherwise, I look forward to seeing the next iteration!

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,37 @@
+#' @examples
+#' pl <- pipelearner(mtcars)
+#' learn_cvpairs(pl, modelr::crossv_kfold, k = 4)
+#' learn_cvpairs(pl, resamplr::holdout_frac, size = 0.3, K = 5L)
```
Please use modelr functions only in public docs.

Commit:
add arbitrary cross-validation functions

Commit:
add arbitrary cross-validation functions (update per Simon's comments)

Commit:
add arbitrary cross validation functions (updated with Simon's comments)

Merge branch 'cv_pairs_funcs' of https://github.com/wdearden208/pipelearner into cv_pairs_funcs

# Conflicts:
#	tests/testthat/test-learn_cvpairs.R

wdearden commented:
Sorry for the triple commit. I made a minor change to the commit after I had already pushed and tried to amend the commit.

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,71 @@
-learn_cvpairs.default <-  function(pl, k = NULL, n = 1, test = 0.2) {
-  learn_cvpairs.pipelearner(pipelearner(pl), k = k, n = n, test = test)
+learn_cvpairs.default <-  function(pl, .f, ...) {
+  learn_cvpairs.pipelearner(pipelearner(pl), ...)
```
`.f` is missing here

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,71 @@
-  # n random partitions
-  if (is.null(k)) {
-    cv_pairs <- modelr::crossv_mc(pl$data, n = n, test = test, id = ".id")
+learn_cvpairs.pipelearner <- function(pl, .f, ...) {
```
This function has become a bit unwieldy. I am happy to fix it to my liking after merging, but there appear to be some more serious issues. Specifically, the default case throws an error. E.g., try running `pipelearner(mtcars, lm, mpg ~ .)` and you'll get an error straight away.

Either way, for default case, I had something simpler in mind. Like:

```
if (missing(.f)) {
  pl$cv_pairs <- crossv_mc(pl$data, n = 1L, test = 0.2, id = ".id")
  return (pl)
}
```

See if this helps with errors.

drsimonj commented on R/learn_cvpairs.R:
```diff
@@ -1,37 +1,71 @@
-  pl$cv_pairs <- cv_pairs
+  # Test object returned by .f
+  assertthat::assert_that(all(c("test", "train", ".id") %in% names(cv)))
+  all_resample <- function(x) {
```
Might as well remove this function. In its place, use...

```
all(purrr::map_lgl(cv$train, inherits, "resample"))
all(purrr::map_lgl(cv$test, inherits, "resample"))
```

Commit:
simply default if .f is missing in cross validation

Commit:
change default value of .f to NULL

drsimonj committed:
Remove test depending on unimported package

One test depended on the resamplr package that is not imported by the package. Hence, it has been removed.

drsimonj committed:
Minor adjustments to learn_cvpairs

drsimonj commented:
I made few tweaks and have now merged this branch. Congrats and thanks on this valuable contribution 👍 

I've also added a [NEWS](https://github.com/drsimonj/pipelearner/blob/master/NEWS.md) doc mentioning this breaking change and thanking you.
